tools/scripts/uconfig_vars_test.py

```diff
@@ -0,0 +1,219 @@
+
+    Returns:
+      List of all UCONFIG_NO_XXX flags found in uconfig.h, initialized test
+      result structure.
+    """
+
+    uconfig_no_flags_all = []
+    test_results = {}
+    uconfig_no_regex = r'UCONFIG_NO_[A-Z_]*'
+
+    # Collect all distinct occurences of UCONFIG_NO_XXX matches in uconfig.h.
+    for uconfig_no_flag in re.finditer(uconfig_no_regex, uconfig_file):
+        if uconfig_no_flag.group(0) not in uconfig_no_flags_all:
+            uconfig_no_flags_all.append(uconfig_no_flag.group(0))
+
+    # All UCONFIG_NO_XXX flags found in uconfig.h come in form of a guarded
+    # definition. Verify the existence, report error if not found.
+    for uconfig_no_flag in uconfig_no_flags_all:
+        uconfig_no_def_regex = r'(?m)#ifndef %s\n#\s+define %s\s+0\n#endif$' % (
+        uconfig_no_flag, uconfig_no_flag)
+        uconfig_no_def_match = re.search(uconfig_no_def_regex, uconfig_file)
+        if not uconfig_no_def_match:
+            print('No definition for flag %s found!\n' % uconfig_no_flag)
+            sys.exit(1)
+
+    test_results = {f: {'unit_test': False, 'hdr_test': False} for f in uconfig_no_flags_all}
+    test_results['all_flags'] = {'unit_test': False, 'hdr_test' : False}
+
+    return uconfig_no_flags_all, test_results
+
+def BuildAllFlags(uconfig_no_list):
+    """Builds sequence of -Dflag=1 with each flag from the list."""
+
+    flag_list = ['-D' + uconfig_no for uconfig_no in uconfig_no_list]
+    flag_value_list = [flag + '=1' for flag in flag_list]
+
+    return ' '.join(flag_value_list)
+
+def RunUnitTests(uconfig_no_list, test_results):
+    """Iterates over all flags, sets each individually during ICU configuration
+       and executes the ICU4C unit tests.
+
+    Args:
+      uconfig_no_list: list of all UCONFIG_NO_XXX flags to test with.
+      test_results: dictionary to record test run results.
+
+    Returns:
+      test_results: updated test result entries.
+    """
+
+    for uconfig_no in uconfig_no_list:
+        _, exit_code = RunCmd(
+                './runConfigureICU Linux CPPFLAGS=\"-D%s=1"' % uconfig_no)
+        if exit_code != 0:
+            print('ICU4C configuration for flag %s failed' % uconfig_no)
+            sys.exit(1)
+        print('Running unit tests with %s set to 1.' % uconfig_no)
+        _, exit_code = RunCmd('make -j2 check')
+        test_results[uconfig_no]['unit_test'] = (exit_code == 0)
+        RunCmd('make clean')
+
+    # Configure ICU with all UCONFIG_NO_XXX flags set to 1 and execute
+    # the ICU4C unit tests.
+    all_unit_test_config_no = BuildAllFlags(uconfig_no_list)
+    _, exit_code = RunCmd(
+        'CPPFLAGS=\"%s\" ./runConfigureICU Linux' % all_unit_test_config_no)
+    if exit_code != 0:
+        print('ICU configuration with all flags set failed')
+        sys.exit(1)
+    print('Running unit tests with all flags set to 1.')
+    _, exit_code = RunCmd('make -j2 check')
+    test_results['all_flags']['unit_test'] = (exit_code == 0)
+
```

Comment:
do another `make clean` here?

Reply:
Added another make clean.
Not really needed I think but sets the client back to a defined state.